<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Koji Wong" />

<meta name="date" content="2025-02-11" />

<title>Assignment 1: Cleaning, Normalizing, and Mapping RNA-Seq Dataset From GEO</title>

<script src="A1_Report_files/header-attrs-2.29/header-attrs.js"></script>
<script src="A1_Report_files/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="A1_Report_files/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="A1_Report_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="A1_Report_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="A1_Report_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="A1_Report_files/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="A1_Report_files/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="A1_Report_files/tocify-1.9.1/jquery.tocify.js"></script>
<script src="A1_Report_files/navigation-1.1/tabsets.js"></script>
<script src="A1_Report_files/navigation-1.1/codefolding.js"></script>
<link href="A1_Report_files/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="A1_Report_files/pagedtable-1.1/js/pagedtable.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>



<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  background-color: #f8f8f8; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #204a87; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #8f5902; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #204a87; font-weight: bold; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #ce5c00; font-weight: bold; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */

.sourceCode .row {
  width: 100%;
}
.sourceCode {
  overflow-x: auto;
}
.code-folding-btn {
  margin-right: -30px;
}
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">
/* for pandoc --citeproc since 2.11 */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Assignment 1: Cleaning, Normalizing, and
Mapping RNA-Seq Dataset From GEO</h1>
<h4 class="author">Koji Wong</h4>
<h4 class="date">02/11/2025</h4>

</div>


<div id="introduction" class="section level2" number="1">
<h2><span class="header-section-number">1</span> Introduction</h2>
<p>The dataset I chose for my assignment is titled “Elevated GRHL2
imparts plasticity in estrogen receptor positive breast cancer cells
[RNA-seq]” with the GEO accession GSE272061 where the dataset can be
obtained <a
href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE272061">here</a>.
This dataset is associated with the 2024 paper by Zheng et al. titled
“Elevated GRHL2 Imparts Plasticity in ER-Positive Breast Cancer Cells”
which can be obtained <a
href="https://www.mdpi.com/2072-6694/16/16/2906">here</a>.</p>
<p>The authors found that “high levels of the transcription factor
Grainyhead-like protein 2 (GRHL2) contribute to worse outcomes for
patients with breast cancer tumors that express estrogen receptor (ER)”
<span class="citation">(Zheng et al. 2024)</span>. The authors
investigated the effects of GRHL2 by creating an overexpression model in
MCF7 cells in which GRHL2-GFP fusion protein was produced upon induction
with doxycycline. RNA-seq was then performed on GFP-positive,
GFP-negative, and negative control cells which were not treated with dox
to give us the raw count values we will be analyzing.</p>
<p>The dataset consists of raw counts for 5 samples with duplicates
corresponding to the three different conditions, giving us a total of 15
samples. The three conditions are GFP-positive, GFP-negative, and
negative control.</p>
</div>
<div id="getting-started" class="section level2" number="2">
<h2><span class="header-section-number">2</span> Getting Started</h2>
<div id="installing-dependencies" class="section level3" number="2.1">
<h3><span class="header-section-number">2.1</span> Installing
dependencies</h3>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="co"># install bioconductor</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a>  <span class="fu">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(<span class="st">&quot;tools&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a>  <span class="fu">install.packages</span>(<span class="st">&quot;tools&quot;</span>)</span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a></span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a><span class="co"># install GEOquery, edgeR, data.table, biomaRt, knitr</span></span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;GEOquery&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a>  BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">&quot;GEOquery&quot;</span>)</span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a></span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;edgeR&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a>  BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">&quot;edgeR&quot;</span>)</span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a></span>
<span id="cb1-14"><a href="#cb1-14" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;data.table&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb1-15"><a href="#cb1-15" tabindex="-1"></a>  BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">&quot;data.table&quot;</span>)</span>
<span id="cb1-16"><a href="#cb1-16" tabindex="-1"></a></span>
<span id="cb1-17"><a href="#cb1-17" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;biomaRt&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb1-18"><a href="#cb1-18" tabindex="-1"></a>  BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">&quot;biomaRt&quot;</span>)</span>
<span id="cb1-19"><a href="#cb1-19" tabindex="-1"></a></span>
<span id="cb1-20"><a href="#cb1-20" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;knitr&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb1-21"><a href="#cb1-21" tabindex="-1"></a>  BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">&quot;knitr&quot;</span>)</span></code></pre></div>
</div>
<div id="looking-at-our-dataset" class="section level3" number="2.2">
<h3><span class="header-section-number">2.2</span> Looking at our
dataset</h3>
<p>We obtain the data directly from GEO using our accession (GSE272061)
and the GEOquery package’s getGEO function.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="co"># getting the data from GEO </span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>data_set_geoid <span class="ot">&lt;-</span> <span class="st">&quot;GSE272061&quot;</span></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a>gse <span class="ot">&lt;-</span> GEOquery<span class="sc">::</span><span class="fu">getGEO</span>(data_set_geoid, <span class="at">GSEMatrix=</span><span class="cn">FALSE</span>)</span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="co"># looking at metadata of dataset</span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a>current_gpl <span class="ot">&lt;-</span> <span class="fu">names</span>(GEOquery<span class="sc">::</span><span class="fu">GPLList</span>(gse))[<span class="dv">1</span>]</span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a>current_gpl_info <span class="ot">&lt;-</span> GEOquery<span class="sc">::</span><span class="fu">Meta</span>(GEOquery<span class="sc">::</span><span class="fu">getGEO</span>(current_gpl))</span></code></pre></div>
<p>Then we can print out some of the metadata about our dataset we
obtained.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="fu">cat</span>(</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>  <span class="fu">c</span>(</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>    <span class="fu">sprintf</span>(<span class="st">&quot;Platform title: %s&quot;</span>, current_gpl_info<span class="sc">$</span>title),</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>    <span class="fu">sprintf</span>(<span class="st">&quot;GEO accession: %s&quot;</span>, current_gpl_info<span class="sc">$</span>geo_accession),</span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>    <span class="fu">sprintf</span>(<span class="st">&quot;Organism: %s&quot;</span>, current_gpl_info<span class="sc">$</span>organism),</span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a>    <span class="fu">sprintf</span>(<span class="st">&quot;Submitted on: %s&quot;</span>, current_gpl_info<span class="sc">$</span>submission_date),</span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a>    <span class="fu">sprintf</span>(<span class="st">&quot;Last updated on: %s&quot;</span>, current_gpl_info<span class="sc">$</span>last_update_date),</span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a>    <span class="fu">sprintf</span>(<span class="st">&quot;Total datasets: %s&quot;</span>, <span class="fu">length</span>(current_gpl_info<span class="sc">$</span>series_id)),</span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a>    <span class="fu">sprintf</span>(<span class="st">&quot;Total samples: %s&quot;</span>, <span class="fu">length</span>(current_gpl_info<span class="sc">$</span>sample_id))</span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a>  ),</span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a>  <span class="at">sep=</span><span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span></span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a>)</span></code></pre></div>
<pre><code>## Platform title: Illumina NovaSeq 6000 (Homo sapiens)
## GEO accession: GPL24676
## Organism: Homo sapiens
## Submitted on: Mar 02 2018
## Last updated on: Nov 05 2018
## Total datasets: 18671
## Total samples: 963881</code></pre>
</div>
</div>
<div id="getting-expression-data" class="section level2" number="3">
<h2><span class="header-section-number">3</span> Getting Expression
Data</h2>
<p>The next step is to download the supplementary files from GEO. In my
case, I only have one supplementary file which allows me to assume I can
just index the first value. I specify where the download will go in the
variable <code>download_dir</code> and download the file only if it
doesn’t exist in that directory already.</p>
<div id="installing-the-dataset" class="section level3" number="3.1">
<h3><span class="header-section-number">3.1</span> Installing the
dataset</h3>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="co"># Getting expression data</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>sfilenames <span class="ot">=</span> GEOquery<span class="sc">::</span><span class="fu">getGEOSuppFiles</span>(data_set_geoid, <span class="at">fetch_files =</span> <span class="cn">FALSE</span>)</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="co"># only have one file to download from GEO, choose the first</span></span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a>raw_data_file <span class="ot">&lt;-</span> sfilenames<span class="sc">$</span>fname[<span class="dv">1</span>]</span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a><span class="co"># file directory to store data in</span></span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a>download_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="fu">getwd</span>(), <span class="st">&quot;data&quot;</span>)</span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">dir.exists</span>(download_dir)) {</span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a>  <span class="fu">dir.create</span>(download_dir)</span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a>}</span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a>file_path <span class="ot">&lt;-</span> <span class="fu">file.path</span>(download_dir, data_set_geoid, raw_data_file)</span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(file_path)) {</span>
<span id="cb5-12"><a href="#cb5-12" tabindex="-1"></a>  GEOquery<span class="sc">::</span><span class="fu">getGEOSuppFiles</span>(</span>
<span id="cb5-13"><a href="#cb5-13" tabindex="-1"></a>    data_set_geoid,</span>
<span id="cb5-14"><a href="#cb5-14" tabindex="-1"></a>    <span class="at">filter_regex =</span> raw_data_file,</span>
<span id="cb5-15"><a href="#cb5-15" tabindex="-1"></a>    <span class="at">baseDir =</span> download_dir,</span>
<span id="cb5-16"><a href="#cb5-16" tabindex="-1"></a>    <span class="at">fetch_files =</span> <span class="cn">TRUE</span></span>
<span id="cb5-17"><a href="#cb5-17" tabindex="-1"></a>  )</span>
<span id="cb5-18"><a href="#cb5-18" tabindex="-1"></a>}</span></code></pre></div>
</div>
<div id="decompressing-the-data" class="section level3" number="3.2">
<h3><span class="header-section-number">3.2</span> Decompressing the
data</h3>
<p>My files were downloaded as a compressed .tar file so I decompressed
it using the builtin <code>tar()</code> function and put it in a folder
called <code>extracted</code></p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="co"># decompressing files</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>extracted_dir <span class="ot">=</span> <span class="fu">file.path</span>(download_dir, <span class="st">&quot;extracted&quot;</span>)</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a>download_path <span class="ot">=</span> <span class="fu">file.path</span>(download_dir, <span class="st">&quot;GSE272061&quot;</span>, raw_data_file)</span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a><span class="fu">untar</span>(download_path, <span class="at">exdir=</span>extracted_dir)</span></code></pre></div>
<p>Hooray, we can access the extracted files now, but all the count data
are in 15 separate files. Let’s combine these together into a convenient
data frame. Along the way, we do some simple error checking, and get the
files which we know are in .gz format.</p>
</div>
<div id="making-our-count-matrix-as-a-data-frame" class="section level3"
number="3.3">
<h3><span class="header-section-number">3.3</span> Making our count
matrix as a data frame</h3>
<p>We want to extract the last column of each data file which contains
our count data. We also label the columns corresponding to the sample
number (eg. p8) to the GFP condition (eg. GFP-positive or negcontrol).
We do this by extracting the condition and sample number from the file
name. Later, I realized it would have been easier to access it by
<code>gse</code>.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="co"># We have 15 files, </span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a><span class="co"># List all .gz files in the directory</span></span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a>gz_files <span class="ot">&lt;-</span> <span class="fu">list.files</span>(extracted_dir, <span class="at">pattern =</span> <span class="st">&quot;</span><span class="sc">\\</span><span class="st">.gz$&quot;</span>, <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a><span class="co"># Initialize a list to store the last columns</span></span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a>last_columns <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a>row_names <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a><span class="co"># Loop through each .gz file</span></span>
<span id="cb7-8"><a href="#cb7-8" tabindex="-1"></a><span class="cf">for</span> (file <span class="cf">in</span> gz_files) {</span>
<span id="cb7-9"><a href="#cb7-9" tabindex="-1"></a>  <span class="co"># Read the .gz file</span></span>
<span id="cb7-10"><a href="#cb7-10" tabindex="-1"></a>  data <span class="ot">&lt;-</span> <span class="fu">tryCatch</span>(</span>
<span id="cb7-11"><a href="#cb7-11" tabindex="-1"></a>    {</span>
<span id="cb7-12"><a href="#cb7-12" tabindex="-1"></a>      <span class="fu">read.table</span>(<span class="fu">gzfile</span>(file), <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>, <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span>
<span id="cb7-13"><a href="#cb7-13" tabindex="-1"></a>    },</span>
<span id="cb7-14"><a href="#cb7-14" tabindex="-1"></a>    <span class="at">error =</span> <span class="cf">function</span>(e) {</span>
<span id="cb7-15"><a href="#cb7-15" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">&quot;Error reading file:&quot;</span>, file, <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>)</span>
<span id="cb7-16"><a href="#cb7-16" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">&quot;Error message:&quot;</span>, e<span class="sc">$</span>message, <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>)</span>
<span id="cb7-17"><a href="#cb7-17" tabindex="-1"></a>      <span class="fu">return</span>(<span class="cn">NULL</span>)</span>
<span id="cb7-18"><a href="#cb7-18" tabindex="-1"></a>    }</span>
<span id="cb7-19"><a href="#cb7-19" tabindex="-1"></a>  )</span>
<span id="cb7-20"><a href="#cb7-20" tabindex="-1"></a>  <span class="co"># Skip if the file couldn&#39;t be read</span></span>
<span id="cb7-21"><a href="#cb7-21" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">is.null</span>(data)) {</span>
<span id="cb7-22"><a href="#cb7-22" tabindex="-1"></a>    <span class="cf">next</span></span>
<span id="cb7-23"><a href="#cb7-23" tabindex="-1"></a>  }</span>
<span id="cb7-24"><a href="#cb7-24" tabindex="-1"></a>  <span class="co"># Check if the data has at least one column</span></span>
<span id="cb7-25"><a href="#cb7-25" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">ncol</span>(data) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb7-26"><a href="#cb7-26" tabindex="-1"></a>    <span class="co"># Extract the last column</span></span>
<span id="cb7-27"><a href="#cb7-27" tabindex="-1"></a>    last_column <span class="ot">&lt;-</span> data[, <span class="fu">ncol</span>(data)]</span>
<span id="cb7-28"><a href="#cb7-28" tabindex="-1"></a>    </span>
<span id="cb7-29"><a href="#cb7-29" tabindex="-1"></a>    <span class="co"># Extract the file name without path and extension</span></span>
<span id="cb7-30"><a href="#cb7-30" tabindex="-1"></a>    file_name <span class="ot">&lt;-</span> tools<span class="sc">::</span><span class="fu">file_path_sans_ext</span>(<span class="fu">basename</span>(file))</span>
<span id="cb7-31"><a href="#cb7-31" tabindex="-1"></a>    </span>
<span id="cb7-32"><a href="#cb7-32" tabindex="-1"></a>    <span class="co"># Clean up the file name to match the p{number}-{condition} format</span></span>
<span id="cb7-33"><a href="#cb7-33" tabindex="-1"></a>    <span class="co"># Eg: &quot;GSM123_p8-GFPnegative.counts&quot; -&gt; &quot;p8-GFPnegative&quot;</span></span>
<span id="cb7-34"><a href="#cb7-34" tabindex="-1"></a>    col_name <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">&quot;^[^_]*_&quot;</span>, <span class="st">&quot;&quot;</span>, file_name)</span>
<span id="cb7-35"><a href="#cb7-35" tabindex="-1"></a>    col_name <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">&quot;</span><span class="sc">\\</span><span class="st">.counts</span><span class="sc">\\</span><span class="st">.txt$&quot;</span>, <span class="st">&quot;&quot;</span>, col_name)</span>
<span id="cb7-36"><a href="#cb7-36" tabindex="-1"></a>    </span>
<span id="cb7-37"><a href="#cb7-37" tabindex="-1"></a>    <span class="co"># Store the last column in the list, using the cleaned-up name as the key</span></span>
<span id="cb7-38"><a href="#cb7-38" tabindex="-1"></a>    last_columns[[col_name]] <span class="ot">&lt;-</span> last_column</span>
<span id="cb7-39"><a href="#cb7-39" tabindex="-1"></a>    </span>
<span id="cb7-40"><a href="#cb7-40" tabindex="-1"></a>    <span class="co"># Store the row names (only once, assuming all files have the same row names)</span></span>
<span id="cb7-41"><a href="#cb7-41" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">is.null</span>(row_names)) {</span>
<span id="cb7-42"><a href="#cb7-42" tabindex="-1"></a>      row_names <span class="ot">&lt;-</span> data[, <span class="st">&quot;Geneid&quot;</span>]</span>
<span id="cb7-43"><a href="#cb7-43" tabindex="-1"></a>    }</span>
<span id="cb7-44"><a href="#cb7-44" tabindex="-1"></a>    </span>
<span id="cb7-45"><a href="#cb7-45" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb7-46"><a href="#cb7-46" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">&quot;File has no columns:&quot;</span>, file, <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>)</span>
<span id="cb7-47"><a href="#cb7-47" tabindex="-1"></a>  }</span>
<span id="cb7-48"><a href="#cb7-48" tabindex="-1"></a>}</span></code></pre></div>
<p>Lastly, we convert the list of our last columns to a data frame and
we set the row names to be the genes.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a>max_length <span class="ot">&lt;-</span> <span class="fu">max</span>(<span class="fu">sapply</span>(last_columns, length))</span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>last_columns <span class="ot">&lt;-</span> <span class="fu">lapply</span>(last_columns, <span class="cf">function</span>(col) {</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">length</span>(col) <span class="sc">&lt;</span> max_length) {</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a>    <span class="fu">c</span>(col, <span class="fu">rep</span>(<span class="cn">NA</span>, max_length <span class="sc">-</span> <span class="fu">length</span>(col)))</span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a>    col</span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a>  }</span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a>})</span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a>count_matrix <span class="ot">&lt;-</span> <span class="fu">do.call</span>(cbind, last_columns)</span>
<span id="cb8-10"><a href="#cb8-10" tabindex="-1"></a><span class="co"># Convert to a data frame</span></span>
<span id="cb8-11"><a href="#cb8-11" tabindex="-1"></a>count_matrix <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(count_matrix)</span>
<span id="cb8-12"><a href="#cb8-12" tabindex="-1"></a><span class="co"># Add row names to the data frame</span></span>
<span id="cb8-13"><a href="#cb8-13" tabindex="-1"></a><span class="fu">rownames</span>(count_matrix) <span class="ot">&lt;-</span> row_names</span></code></pre></div>
</div>
<div id="looking-at-the-raw-data" class="section level3" number="3.4">
<h3><span class="header-section-number">3.4</span> Looking at the raw
data</h3>
<p>We can view the number of rows (or genes) and the number of columns
(or conditions)</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="co"># Printing out the data</span></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a><span class="fu">cat</span>(</span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a>  <span class="fu">c</span>(</span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a>    <span class="fu">sprintf</span>(<span class="st">&quot;Summary of Count Matrix&quot;</span>),</span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a>    <span class="fu">sprintf</span>(<span class="st">&quot;number of genes: %d&quot;</span>, <span class="fu">nrow</span>(count_matrix)),</span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a>    <span class="fu">sprintf</span>(<span class="st">&quot;number of conditions: %d&quot;</span>, <span class="fu">ncol</span>(count_matrix))</span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a>  ),</span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a>  <span class="at">sep=</span><span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span></span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a>)</span></code></pre></div>
<pre><code>## Summary of Count Matrix
## number of genes: 57500
## number of conditions: 15</code></pre>
</div>
<div id="looking-at-the-metadata" class="section level3" number="3.5">
<h3><span class="header-section-number">3.5</span> Looking at the
metadata</h3>
<p>We can view the metadata for each of our samples, including the
title, accession, tissue, and treatment.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a>samples <span class="ot">&lt;-</span> gse<span class="sc">@</span>gsms</span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>samples_type <span class="ot">&lt;-</span> <span class="fu">do.call</span>(rbind,</span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a>                        <span class="fu">lapply</span>(samples,</span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a>                               <span class="at">FUN=</span><span class="cf">function</span>(x) {</span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a>                                 <span class="fu">c</span>(x<span class="sc">@</span>header<span class="sc">$</span>title,</span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a>                                   x<span class="sc">@</span>header<span class="sc">$</span>characetristics_ch1)</span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a>                               }))</span>
<span id="cb11-8"><a href="#cb11-8" tabindex="-1"></a></span>
<span id="cb11-9"><a href="#cb11-9" tabindex="-1"></a><span class="co"># Make a table of metadata</span></span>
<span id="cb11-10"><a href="#cb11-10" tabindex="-1"></a>metadata_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb11-11"><a href="#cb11-11" tabindex="-1"></a>  <span class="at">title =</span> <span class="fu">character</span>(),</span>
<span id="cb11-12"><a href="#cb11-12" tabindex="-1"></a>  <span class="at">accession =</span> <span class="fu">character</span>(),</span>
<span id="cb11-13"><a href="#cb11-13" tabindex="-1"></a>  <span class="at">tissue =</span> <span class="fu">character</span>(),</span>
<span id="cb11-14"><a href="#cb11-14" tabindex="-1"></a>  <span class="at">cell_line =</span> <span class="fu">character</span>(),</span>
<span id="cb11-15"><a href="#cb11-15" tabindex="-1"></a>  <span class="at">treatment =</span> <span class="fu">character</span>(),</span>
<span id="cb11-16"><a href="#cb11-16" tabindex="-1"></a>  <span class="at">sorting =</span> <span class="fu">character</span>(),</span>
<span id="cb11-17"><a href="#cb11-17" tabindex="-1"></a>  <span class="at">library_strategy =</span> <span class="fu">character</span>()</span>
<span id="cb11-18"><a href="#cb11-18" tabindex="-1"></a>)</span>
<span id="cb11-19"><a href="#cb11-19" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(samples)) {</span>
<span id="cb11-20"><a href="#cb11-20" tabindex="-1"></a>  <span class="co"># Save header variable for easy reference</span></span>
<span id="cb11-21"><a href="#cb11-21" tabindex="-1"></a>  header <span class="ot">&lt;-</span> samples[[i]]<span class="sc">@</span>header</span>
<span id="cb11-22"><a href="#cb11-22" tabindex="-1"></a>  <span class="co"># Get list of values we need</span></span>
<span id="cb11-23"><a href="#cb11-23" tabindex="-1"></a>  title <span class="ot">&lt;-</span> header<span class="sc">$</span>title</span>
<span id="cb11-24"><a href="#cb11-24" tabindex="-1"></a>  accession <span class="ot">&lt;-</span> header<span class="sc">$</span>geo_accession</span>
<span id="cb11-25"><a href="#cb11-25" tabindex="-1"></a>  tissue <span class="ot">&lt;-</span> header<span class="sc">$</span>characteristics_ch1[<span class="dv">1</span>]</span>
<span id="cb11-26"><a href="#cb11-26" tabindex="-1"></a>  cell_line <span class="ot">&lt;-</span> header<span class="sc">$</span>characteristics_ch1[<span class="dv">2</span>]</span>
<span id="cb11-27"><a href="#cb11-27" tabindex="-1"></a>  treatment <span class="ot">&lt;-</span> header<span class="sc">$</span>characteristics_ch1[<span class="dv">3</span>]</span>
<span id="cb11-28"><a href="#cb11-28" tabindex="-1"></a>  sorting <span class="ot">&lt;-</span> header<span class="sc">$</span>characteristics_ch1[<span class="dv">4</span>]</span>
<span id="cb11-29"><a href="#cb11-29" tabindex="-1"></a>  library_strategy <span class="ot">&lt;-</span> header<span class="sc">$</span>library_strategy</span>
<span id="cb11-30"><a href="#cb11-30" tabindex="-1"></a>  </span>
<span id="cb11-31"><a href="#cb11-31" tabindex="-1"></a>  <span class="co"># Create a temporary df to append to our current one</span></span>
<span id="cb11-32"><a href="#cb11-32" tabindex="-1"></a>  tmp <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb11-33"><a href="#cb11-33" tabindex="-1"></a>    <span class="at">title =</span> title,</span>
<span id="cb11-34"><a href="#cb11-34" tabindex="-1"></a>    <span class="at">accession =</span> accession,</span>
<span id="cb11-35"><a href="#cb11-35" tabindex="-1"></a>    <span class="at">tissue =</span> tissue,</span>
<span id="cb11-36"><a href="#cb11-36" tabindex="-1"></a>    <span class="at">cell_line =</span> cell_line,</span>
<span id="cb11-37"><a href="#cb11-37" tabindex="-1"></a>    <span class="at">treatment =</span> treatment,</span>
<span id="cb11-38"><a href="#cb11-38" tabindex="-1"></a>    <span class="at">sorting =</span> sorting,</span>
<span id="cb11-39"><a href="#cb11-39" tabindex="-1"></a>    <span class="at">library_strategy =</span> library_strategy</span>
<span id="cb11-40"><a href="#cb11-40" tabindex="-1"></a>  )</span>
<span id="cb11-41"><a href="#cb11-41" tabindex="-1"></a>  </span>
<span id="cb11-42"><a href="#cb11-42" tabindex="-1"></a>  <span class="co"># Bind dataframe and set as old one</span></span>
<span id="cb11-43"><a href="#cb11-43" tabindex="-1"></a>  metadata_df <span class="ot">&lt;-</span> <span class="fu">rbind</span>(metadata_df, tmp)</span>
<span id="cb11-44"><a href="#cb11-44" tabindex="-1"></a>}</span></code></pre></div>
<p>Let’s look at the metadata</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="fu">head</span>(metadata_df, <span class="dv">15</span>)</span></code></pre></div>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["title"],"name":[1],"type":["chr"],"align":["left"]},{"label":["accession"],"name":[2],"type":["chr"],"align":["left"]},{"label":["tissue"],"name":[3],"type":["chr"],"align":["left"]},{"label":["cell_line"],"name":[4],"type":["chr"],"align":["left"]},{"label":["treatment"],"name":[5],"type":["chr"],"align":["left"]},{"label":["sorting"],"name":[6],"type":["chr"],"align":["left"]},{"label":["library_strategy"],"name":[7],"type":["chr"],"align":["left"]}],"data":[{"1":"p8 GFP negative (Biol rep 1)","2":"GSM8392266","3":"tissue: Breast cancer","4":"cell line: MCF7 RTA16 tet-ind. GRHL2-GFP cl. 62 (OE)","5":"treatment: 1 μg/mL Dox for 72 hours","6":"sorting: GFP negative","7":"RNA-Seq","_rn_":"1"},{"1":"p8 GFP positive (Biol rep 1)","2":"GSM8392267","3":"tissue: Breast cancer","4":"cell line: MCF7 RTA16 tet-ind. GRHL2-GFP cl. 62 (OE)","5":"treatment: 1 μg/mL Dox for 72 hours","6":"sorting: GFP positive","7":"RNA-Seq","_rn_":"2"},{"1":"p8 neg control (Biol rep 1)","2":"GSM8392268","3":"tissue: Breast cancer","4":"cell line: MCF7 RTA16 tet-ind. GRHL2-GFP cl. 62 (OE)","5":"treatment: vehicle","6":"sorting: N/A","7":"RNA-Seq","_rn_":"3"},{"1":"p9 GFP negative (Biol rep 2)","2":"GSM8392269","3":"tissue: Breast cancer","4":"cell line: MCF7 RTA16 tet-ind. GRHL2-GFP cl. 62 (OE)","5":"treatment: 1 μg/mL Dox for 72 hours","6":"sorting: GFP negative","7":"RNA-Seq","_rn_":"4"},{"1":"p9 GFP positive (Biol rep 2)","2":"GSM8392270","3":"tissue: Breast cancer","4":"cell line: MCF7 RTA16 tet-ind. GRHL2-GFP cl. 62 (OE)","5":"treatment: 1 μg/mL Dox for 72 hours","6":"sorting: GFP positive","7":"RNA-Seq","_rn_":"5"},{"1":"p9 neg control (Biol rep 2)","2":"GSM8392271","3":"tissue: Breast cancer","4":"cell line: MCF7 RTA16 tet-ind. GRHL2-GFP cl. 62 (OE)","5":"treatment: vehicle","6":"sorting: N/A","7":"RNA-Seq","_rn_":"6"},{"1":"p10 GFP negative (Biol rep 3)","2":"GSM8392272","3":"tissue: Breast cancer","4":"cell line: MCF7 RTA16 tet-ind. GRHL2-GFP cl. 62 (OE)","5":"treatment: 1 μg/mL Dox for 72 hours","6":"sorting: GFP negative","7":"RNA-Seq","_rn_":"7"},{"1":"p10 GFP positive (Biol rep 3)","2":"GSM8392273","3":"tissue: Breast cancer","4":"cell line: MCF7 RTA16 tet-ind. GRHL2-GFP cl. 62 (OE)","5":"treatment: 1 μg/mL Dox for 72 hours","6":"sorting: GFP positive","7":"RNA-Seq","_rn_":"8"},{"1":"p10 neg control (Biol rep 3)","2":"GSM8392274","3":"tissue: Breast cancer","4":"cell line: MCF7 RTA16 tet-ind. GRHL2-GFP cl. 62 (OE)","5":"treatment: vehicle","6":"sorting: N/A","7":"RNA-Seq","_rn_":"9"},{"1":"p11 GFP negative (Biol rep 4)","2":"GSM8392275","3":"tissue: Breast cancer","4":"cell line: MCF7 RTA16 tet-ind. GRHL2-GFP cl. 62 (OE)","5":"treatment: 1 μg/mL Dox for 72 hours","6":"sorting: GFP negative","7":"RNA-Seq","_rn_":"10"},{"1":"p11 GFP positive (Biol rep 4)","2":"GSM8392276","3":"tissue: Breast cancer","4":"cell line: MCF7 RTA16 tet-ind. GRHL2-GFP cl. 62 (OE)","5":"treatment: 1 μg/mL Dox for 72 hours","6":"sorting: GFP positive","7":"RNA-Seq","_rn_":"11"},{"1":"p11 neg control (Biol rep 4)","2":"GSM8392277","3":"tissue: Breast cancer","4":"cell line: MCF7 RTA16 tet-ind. GRHL2-GFP cl. 62 (OE)","5":"treatment: vehicle","6":"sorting: N/A","7":"RNA-Seq","_rn_":"12"},{"1":"p12 GFP negative (Biol rep 5)","2":"GSM8392278","3":"tissue: Breast cancer","4":"cell line: MCF7 RTA16 tet-ind. GRHL2-GFP cl. 62 (OE)","5":"treatment: 1 μg/mL Dox for 72 hours","6":"sorting: GFP negative","7":"RNA-Seq","_rn_":"13"},{"1":"p12 GFP positive (Biol rep 5)","2":"GSM8392279","3":"tissue: Breast cancer","4":"cell line: MCF7 RTA16 tet-ind. GRHL2-GFP cl. 62 (OE)","5":"treatment: 1 μg/mL Dox for 72 hours","6":"sorting: GFP positive","7":"RNA-Seq","_rn_":"14"},{"1":"p12 neg control (Biol rep 5)","2":"GSM8392280","3":"tissue: Breast cancer","4":"cell line: MCF7 RTA16 tet-ind. GRHL2-GFP cl. 62 (OE)","5":"treatment: vehicle","6":"sorting: N/A","7":"RNA-Seq","_rn_":"15"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<div id="filtering" class="section level2" number="4">
<h2><span class="header-section-number">4</span> Filtering</h2>
<p>Let’s look at the first couple of rows</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="co"># Check the first 10 rows</span></span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a><span class="fu">head</span>(count_matrix, <span class="dv">10</span>)</span></code></pre></div>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["p8-GFPnegative"],"name":[1],"type":["int"],"align":["right"]},{"label":["p8-GFPpositive"],"name":[2],"type":["int"],"align":["right"]},{"label":["p8-negcontrol"],"name":[3],"type":["int"],"align":["right"]},{"label":["p9-GFPnegative"],"name":[4],"type":["int"],"align":["right"]},{"label":["p9-GFPpositive"],"name":[5],"type":["int"],"align":["right"]},{"label":["p9-negcontrol"],"name":[6],"type":["int"],"align":["right"]},{"label":["p10-GFP"],"name":[7],"type":["int"],"align":["right"]},{"label":["p10-GFPpositive"],"name":[8],"type":["int"],"align":["right"]},{"label":["p10-negcontrol"],"name":[9],"type":["int"],"align":["right"]},{"label":["p11-GFP"],"name":[10],"type":["int"],"align":["right"]},{"label":["p11-GFPpositive"],"name":[11],"type":["int"],"align":["right"]},{"label":["p11-negcontrol"],"name":[12],"type":["int"],"align":["right"]},{"label":["p12-GFP"],"name":[13],"type":["int"],"align":["right"]},{"label":["p12-GFPpositive"],"name":[14],"type":["int"],"align":["right"]},{"label":["p12-negcontrol"],"name":[15],"type":["int"],"align":["right"]}],"data":[{"1":"0","2":"1","3":"1","4":"1","5":"1","6":"0","7":"0","8":"0","9":"0","10":"0","11":"0","12":"0","13":"0","14":"0","15":"1","_rn_":"ENSG00000223972"},{"1":"45","2":"53","3":"37","4":"38","5":"43","6":"51","7":"18","8":"16","9":"24","10":"30","11":"37","12":"50","13":"26","14":"20","15":"33","_rn_":"ENSG00000227232"},{"1":"4","2":"4","3":"1","4":"4","5":"3","6":"1","7":"1","8":"0","9":"2","10":"0","11":"1","12":"0","13":"0","14":"0","15":"4","_rn_":"ENSG00000278267"},{"1":"0","2":"0","3":"1","4":"0","5":"0","6":"1","7":"0","8":"0","9":"0","10":"0","11":"0","12":"0","13":"0","14":"0","15":"0","_rn_":"ENSG00000243485"},{"1":"0","2":"0","3":"0","4":"0","5":"0","6":"0","7":"0","8":"0","9":"0","10":"0","11":"0","12":"0","13":"0","14":"0","15":"0","_rn_":"ENSG00000237613"},{"1":"0","2":"0","3":"0","4":"0","5":"0","6":"0","7":"0","8":"0","9":"0","10":"0","11":"0","12":"0","13":"0","14":"0","15":"0","_rn_":"ENSG00000268020"},{"1":"0","2":"0","3":"0","4":"0","5":"0","6":"0","7":"0","8":"0","9":"0","10":"0","11":"0","12":"0","13":"0","14":"0","15":"0","_rn_":"ENSG00000240361"},{"1":"0","2":"0","3":"0","4":"0","5":"0","6":"0","7":"0","8":"0","9":"0","10":"0","11":"0","12":"0","13":"0","14":"0","15":"0","_rn_":"ENSG00000186092"},{"1":"4","2":"3","3":"3","4":"1","5":"10","6":"3","7":"0","8":"0","9":"2","10":"1","11":"2","12":"1","13":"0","14":"0","15":"3","_rn_":"ENSG00000238009"},{"1":"0","2":"0","3":"0","4":"0","5":"0","6":"0","7":"0","8":"0","9":"0","10":"0","11":"0","12":"0","13":"0","14":"0","15":"0","_rn_":"ENSG00000239945"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<p>That’s quite a lot of 0’s, it seems like we can clean this up by
removing genes with low total counts.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="co"># Normalizing our data</span></span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a><span class="co"># Eliminate genes with low counts</span></span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a></span>
<span id="cb14-4"><a href="#cb14-4" tabindex="-1"></a>sample_type_dt <span class="ot">&lt;-</span> data.table<span class="sc">::</span><span class="fu">data.table</span>(count_matrix)</span>
<span id="cb14-5"><a href="#cb14-5" tabindex="-1"></a>min_num_samples <span class="ot">&lt;-</span> <span class="dv">16</span></span>
<span id="cb14-6"><a href="#cb14-6" tabindex="-1"></a><span class="co"># Calculate row sums</span></span>
<span id="cb14-7"><a href="#cb14-7" tabindex="-1"></a>row_sums <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(count_matrix)</span>
<span id="cb14-8"><a href="#cb14-8" tabindex="-1"></a><span class="co"># Filter rows where row sums are greater than or equal to 20</span></span>
<span id="cb14-9"><a href="#cb14-9" tabindex="-1"></a>filtered_dataset <span class="ot">&lt;-</span> count_matrix[row_sums <span class="sc">&gt;</span> min_num_samples, ]</span></code></pre></div>
<p>Below we can see the number of genes that were filtered out, just
from low counts.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="fu">cat</span>(<span class="fu">sprintf</span>(<span class="st">&quot;Number of genes before filtering: %s&quot;</span>, <span class="fu">nrow</span>(count_matrix)))</span></code></pre></div>
<pre><code>## Number of genes before filtering: 57500</code></pre>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="fu">cat</span>(<span class="fu">sprintf</span>(<span class="st">&quot;Number of genes after filtering low counts: %s&quot;</span>, <span class="fu">nrow</span>(filtered_dataset)))</span></code></pre></div>
<pre><code>## Number of genes after filtering low counts: 21975</code></pre>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="fu">cat</span>(<span class="fu">sprintf</span>(<span class="st">&quot;Total number of genes filtered out: %s&quot;</span>, <span class="fu">nrow</span>(count_matrix) <span class="sc">-</span> <span class="fu">nrow</span>(filtered_dataset)))</span></code></pre></div>
<pre><code>## Total number of genes filtered out: 35525</code></pre>
<p>We filtered out 35525 genes! This makes sense as we’re working with
samples from breast cancer tissue specifically, so many genes would not
be differentially expressed and present in RNA-seq data.</p>
</div>
<div id="normalization" class="section level2" number="5">
<h2><span class="header-section-number">5</span> Normalization</h2>
<p>The next step after filtering for low counts is to normalize our
data. We do so using the edgeR package which contains functions to help
us normalize. We use Trimmed Means of M-values (TMM) which is based on
the hypothesis that most genes are not differentially expressed. Perfect
for our use case of RNA-seq.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a><span class="co"># Create a DGEList object, make sure we use a matrix</span></span>
<span id="cb21-2"><a href="#cb21-2" tabindex="-1"></a>dge <span class="ot">&lt;-</span> edgeR<span class="sc">::</span><span class="fu">DGEList</span>(<span class="at">counts =</span> <span class="fu">as.matrix</span>(filtered_dataset))</span>
<span id="cb21-3"><a href="#cb21-3" tabindex="-1"></a><span class="co"># Calculate TMM normalization factors</span></span>
<span id="cb21-4"><a href="#cb21-4" tabindex="-1"></a>dge <span class="ot">&lt;-</span> edgeR<span class="sc">::</span><span class="fu">calcNormFactors</span>(dge, <span class="at">method =</span> <span class="st">&quot;TMM&quot;</span>)</span>
<span id="cb21-5"><a href="#cb21-5" tabindex="-1"></a><span class="co"># Get normalized counts per million (log2-transformed)</span></span>
<span id="cb21-6"><a href="#cb21-6" tabindex="-1"></a>normalized_counts <span class="ot">&lt;-</span> edgeR<span class="sc">::</span><span class="fu">cpm</span>(dge, <span class="at">log =</span> <span class="cn">TRUE</span>)</span>
<span id="cb21-7"><a href="#cb21-7" tabindex="-1"></a><span class="co"># Convert to a data frame</span></span>
<span id="cb21-8"><a href="#cb21-8" tabindex="-1"></a>normalized_df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(normalized_counts)</span></code></pre></div>
</div>
<div id="mapping-to-gene-symbols" class="section level2" number="6">
<h2><span class="header-section-number">6</span> Mapping to Gene
Symbols</h2>
<p>The last step of our analysis is to map our ensembl gene IDs to HGNC
symbols. We do so using the biomaRt package. We also save the data as a
.rds file since this is an expensive operation.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a><span class="co"># Mapping our data to gene symbols</span></span>
<span id="cb22-2"><a href="#cb22-2" tabindex="-1"></a>ensembl <span class="ot">&lt;-</span> biomaRt<span class="sc">::</span><span class="fu">useMart</span>(<span class="st">&quot;ensembl&quot;</span>)</span>
<span id="cb22-3"><a href="#cb22-3" tabindex="-1"></a></span>
<span id="cb22-4"><a href="#cb22-4" tabindex="-1"></a>datasets <span class="ot">&lt;-</span> biomaRt<span class="sc">::</span><span class="fu">listDatasets</span>(ensembl)</span>
<span id="cb22-5"><a href="#cb22-5" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(datasets[<span class="fu">grep</span>(datasets<span class="sc">$</span>dataset,</span>
<span id="cb22-6"><a href="#cb22-6" tabindex="-1"></a>                                <span class="at">pattern=</span><span class="st">&quot;sapiens&quot;</span>),]), <span class="at">format=</span><span class="st">&quot;html&quot;</span>)</span></code></pre></div>
<table>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:left;">
dataset
</th>
<th style="text-align:left;">
description
</th>
<th style="text-align:left;">
version
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
80
</td>
<td style="text-align:left;">
hsapiens_gene_ensembl
</td>
<td style="text-align:left;">
Human genes (GRCh38.p14)
</td>
<td style="text-align:left;">
GRCh38.p14
</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a><span class="co"># Use homo sapiens gene ensembl dataset</span></span>
<span id="cb23-2"><a href="#cb23-2" tabindex="-1"></a>ensembl <span class="ot">&lt;-</span> biomaRt<span class="sc">::</span><span class="fu">useDataset</span>(<span class="st">&quot;hsapiens_gene_ensembl&quot;</span>, <span class="at">mart=</span>ensembl)</span>
<span id="cb23-3"><a href="#cb23-3" tabindex="-1"></a><span class="co"># Want to get attributes gene Ids and HGNC symbols</span></span>
<span id="cb23-4"><a href="#cb23-4" tabindex="-1"></a>genes <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">rownames</span>(normalized_df))</span>
<span id="cb23-5"><a href="#cb23-5" tabindex="-1"></a><span class="co"># Remove any version numbers</span></span>
<span id="cb23-6"><a href="#cb23-6" tabindex="-1"></a>genes <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">&quot;</span><span class="sc">\\</span><span class="st">..*&quot;</span>, <span class="st">&quot;&quot;</span>, genes)</span>
<span id="cb23-7"><a href="#cb23-7" tabindex="-1"></a><span class="co"># Check if we have already processed data, very expensive task</span></span>
<span id="cb23-8"><a href="#cb23-8" tabindex="-1"></a>conversion_stash <span class="ot">&lt;-</span> <span class="st">&quot;id_conversion.rds&quot;</span></span>
<span id="cb23-9"><a href="#cb23-9" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">file.exists</span>(conversion_stash)) {</span>
<span id="cb23-10"><a href="#cb23-10" tabindex="-1"></a>  id_conversion <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(conversion_stash)</span>
<span id="cb23-11"><a href="#cb23-11" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb23-12"><a href="#cb23-12" tabindex="-1"></a>  id_conversion <span class="ot">&lt;-</span> biomaRt<span class="sc">::</span><span class="fu">getBM</span>(</span>
<span id="cb23-13"><a href="#cb23-13" tabindex="-1"></a>    <span class="at">attributes =</span> <span class="fu">c</span>(<span class="st">&quot;ensembl_gene_id&quot;</span>, <span class="st">&quot;hgnc_symbol&quot;</span>),</span>
<span id="cb23-14"><a href="#cb23-14" tabindex="-1"></a>    <span class="at">filters =</span> <span class="st">&quot;ensembl_gene_id&quot;</span>,</span>
<span id="cb23-15"><a href="#cb23-15" tabindex="-1"></a>    <span class="at">values =</span> genes,</span>
<span id="cb23-16"><a href="#cb23-16" tabindex="-1"></a>    <span class="at">mart =</span> ensembl</span>
<span id="cb23-17"><a href="#cb23-17" tabindex="-1"></a>  )</span>
<span id="cb23-18"><a href="#cb23-18" tabindex="-1"></a>  <span class="fu">saveRDS</span>(id_conversion, conversion_stash)</span>
<span id="cb23-19"><a href="#cb23-19" tabindex="-1"></a>}</span></code></pre></div>
<p>After that, we merge the gene symbols in with the normalized count
values.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" tabindex="-1"></a>normalized_counts_annot <span class="ot">&lt;-</span> <span class="fu">merge</span>(id_conversion, normalized_df,</span>
<span id="cb24-2"><a href="#cb24-2" tabindex="-1"></a>                                 <span class="at">by.x =</span> <span class="dv">1</span>, <span class="at">by.y =</span> <span class="dv">0</span>, <span class="at">all.y=</span><span class="cn">TRUE</span>)</span></code></pre></div>
<p>We can view how many genes we’ve had before and how many we have
after mapping.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" tabindex="-1"></a>num_mapped_genes <span class="ot">=</span> <span class="fu">length</span>(<span class="fu">which</span>(<span class="fu">rownames</span>(normalized_counts) <span class="sc">%in%</span></span>
<span id="cb25-2"><a href="#cb25-2" tabindex="-1"></a>                                  id_conversion<span class="sc">$</span>ensembl_gene_id))</span>
<span id="cb25-3"><a href="#cb25-3" tabindex="-1"></a><span class="fu">cat</span>(</span>
<span id="cb25-4"><a href="#cb25-4" tabindex="-1"></a>  <span class="fu">c</span>(</span>
<span id="cb25-5"><a href="#cb25-5" tabindex="-1"></a>    <span class="fu">sprintf</span>(<span class="st">&quot;Number of ensembl genes we started with: %d&quot;</span>, <span class="fu">nrow</span>(normalized_df)),</span>
<span id="cb25-6"><a href="#cb25-6" tabindex="-1"></a>    <span class="fu">sprintf</span>(<span class="st">&quot;Number of ensembl genes we were able to map: %d&quot;</span>, num_mapped_genes),</span>
<span id="cb25-7"><a href="#cb25-7" tabindex="-1"></a>    <span class="fu">sprintf</span>(<span class="st">&quot;Number of genes that were not able to be mapped: %d&quot;</span>, <span class="fu">nrow</span>(normalized_df) <span class="sc">-</span> num_mapped_genes)</span>
<span id="cb25-8"><a href="#cb25-8" tabindex="-1"></a>  ),</span>
<span id="cb25-9"><a href="#cb25-9" tabindex="-1"></a>  <span class="at">sep=</span><span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span></span>
<span id="cb25-10"><a href="#cb25-10" tabindex="-1"></a>)</span></code></pre></div>
<pre><code>## Number of ensembl genes we started with: 21975
## Number of ensembl genes we were able to map: 21496
## Number of genes that were not able to be mapped: 479</code></pre>
<p>We see that we’ve mapped most of our genes, only 479 were unable to
be mapped. Let’s also look at the normalized data with the symbols.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(normalized_counts_annot[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>], <span class="at">type=</span><span class="st">&quot;html&quot;</span>)</span></code></pre></div>
<table>
<colgroup>
<col width="22%" />
<col width="16%" />
<col width="20%" />
<col width="20%" />
<col width="19%" />
</colgroup>
<thead>
<tr class="header">
<th align="left">ensembl_gene_id</th>
<th align="left">hgnc_symbol</th>
<th align="right">p8-GFPnegative</th>
<th align="right">p8-GFPpositive</th>
<th align="right">p8-negcontrol</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">ENSG00000000003</td>
<td align="left">TSPAN6</td>
<td align="right">5.112333</td>
<td align="right">4.9590878</td>
<td align="right">5.395526</td>
</tr>
<tr class="even">
<td align="left">ENSG00000000419</td>
<td align="left">DPM1</td>
<td align="right">5.487740</td>
<td align="right">5.4894252</td>
<td align="right">5.693004</td>
</tr>
<tr class="odd">
<td align="left">ENSG00000000457</td>
<td align="left">SCYL3</td>
<td align="right">3.969309</td>
<td align="right">3.5187462</td>
<td align="right">4.013102</td>
</tr>
<tr class="even">
<td align="left">ENSG00000000460</td>
<td align="left">FIRRM</td>
<td align="right">3.670472</td>
<td align="right">3.4970203</td>
<td align="right">3.999067</td>
</tr>
<tr class="odd">
<td align="left">ENSG00000000938</td>
<td align="left">FGR</td>
<td align="right">-3.349927</td>
<td align="right">0.1711314</td>
<td align="right">-2.735221</td>
</tr>
</tbody>
</table>
</div>
<div id="questions" class="section level2" number="7">
<h2><span class="header-section-number">7</span> Questions</h2>
<p><strong>1. Why is the dataset of interest to you?</strong></p>
<p>This dataset was of interest to me as I have always been curious
about RNA-Seq data about cancers, specifically breast cancer.</p>
<p><strong>2. What are the control and test conditions of the
dataset?</strong></p>
<p>The test conditions are either GFP-positive or GFP-negative and are
treated with 1 μg/mL doxycycline hyclate and 10 nM 17β-estradiol for 72
hours, changing the media every 24 hours. The negative control was
treated with vehicle with media changes every 24 hours as well.</p>
<p><strong>3. How many samples in each of the conditions of your
dataset?</strong></p>
<p>For each of the 3 conditions there were 5 samples, giving us 15 total
columns.</p>
<p><strong>4. Were there expression values that were not unique for
specific genes? How did you handle these?</strong></p>
<p>Many expression values had 0 counts across all conditions/samples,
which is understood through the lack of differential expression in these
breast cancer tissues.</p>
<p>These rows were simply excluded as they would create noise in the
normalization process. Specifically, all expression values with count
sums less than 16 were removed. This threshold was chosen in that if
there was a row with 16 counts, then at least one condition has more
than 1 count, since there are 15 rows</p>
<p><strong>5. Were there expression values that could not be mapped to
current HUGO symbols?</strong></p>
<p>Yes, 479 expression values could not be mapped to current HUGO
symbols.</p>
<p><strong>6. Were there any outliers in your dataset? How were they
handled in the originating paper? How many outliers were
removed?</strong></p>
<p>The expression values with less than 16 were removed. In total, there
were 35525 rows which were removed from filtering.</p>
<p><strong>7. How did you handle replicates?</strong></p>
<p>There were no duplicate rows in the initial dataset. There were
however multiple rows pointing to the same HGNC symbol. Since data was
normalized and then mapped it was difficult to come up with a solution
to handle this.</p>
<p><strong>8. What is the final coverage of your dataset?</strong></p>
<p>The initial dataset consisted of 57500 rows. The final dataset has
21496 rows each mapping to a unique HGNC symbol.</p>
</div>
<div id="works-cited" class="section level2 unnumbered">
<h2 class="unnumbered">Works Cited</h2>
<div id="refs" class="references csl-bib-body hanging-indent"
entry-spacing="0">
<div id="ref-datatable" class="csl-entry">
Barrett, Tyson, Matt Dowle, and Arun Srinivasan. 2023. <em>Data.table:
Extension of ‘Data.frame‘</em>. <a
href="https://CRAN.R-project.org/package=data.table">https://CRAN.R-project.org/package=data.table</a>.
</div>
<div id="ref-edger" class="csl-entry">
Chen, Yunshun, Aaron A T Lun, and Gordon K Smyth. 2016. <span>“From
Reads to Genes to Pathways: Differential Expression Analysis of RNA-Seq
Experiments Using Rsubread and the edgeR Quasi-Likelihood
Pipeline.”</span> <em>F1000Research</em> 5: 1438. <a
href="https://doi.org/10.12688/f1000research.8987.2">https://doi.org/10.12688/f1000research.8987.2</a>.
</div>
<div id="ref-geoquery" class="csl-entry">
Davis, Sean, and Paul Meltzer. 2007. <span>“GEOquery: A Bridge Between
the Gene Expression Omnibus (GEO) and BioConductor.”</span>
<em>Bioinformatics</em> 14: 1846–47.
</div>
<div id="ref-biomart" class="csl-entry">
Durinck, Steffen, Yves Moreau, Arek Kasprzyk, Sean Davis, Bart De Moor,
Alvis Brazma, and Wolfgang Huber. 2005. <span>“BioMart and Bioconductor:
A Powerful Link Between Biological Databases and Microarray Data
Analysis.”</span> <em>Bioinformatics</em> 21: 3439–40.
</div>
<div id="ref-biocmanager" class="csl-entry">
Morgan, Martin, and Marcel Ramos. 2023. <em>BiocManager: Access the
Bioconductor Project Package Repository</em>. <a
href="https://CRAN.R-project.org/package=BiocManager">https://CRAN.R-project.org/package=BiocManager</a>.
</div>
<div id="ref-knitr" class="csl-entry">
Xie, Yihui. 2014. <span>“Knitr: A Comprehensive Tool for Reproducible
Research in <span>R</span>.”</span> In <em>Implementing Reproducible
Computational Research</em>, edited by Victoria Stodden, Friedrich
Leisch, and Roger D. Peng. Chapman; Hall/CRC.
</div>
<div id="ref-zheng2024" class="csl-entry">
Zheng, Christy, Kaelyn O. Allen, Tianrui Liu, Natalia M. Solodin, Mark
B. Meyer, Kelley Salem, Phillipos K. Tsourkas, et al. 2024.
<span>“Elevated <span>GRHL2</span> Imparts Plasticity in
<span>ER</span>-Positive Breast Cancer Cells.”</span> <em>Cancers</em>
16 (16): 2906. <a
href="https://doi.org/10.3390/cancers16162906">https://doi.org/10.3390/cancers16162906</a>.
</div>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("show" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
